std/assembly/array.ts

```diff
@@ -476,6 +476,62 @@ export class Array<T> extends ArrayBufferView {
     return this;
   }
 
+  flat(): valueof<T>[] {
+    assert(isArray<T>());
+    // Get the length and data start values
+    let length = this.length_;
+    let selfDataStart = this.dataStart;
+
+    // calculate the end size with an initial pass
+    let size: i32 = 0;
+    for (let i = 0; i < length; i++) {
+      let child: T = load<T>(selfDataStart + (i << alignof<T>()));
+      size += child == null ? 1 : child.length;
+    }
+
+    // calculate the byteLength of the resulting backing ArrayBuffer
+    let byteLength = size << usize(alignof<valueof<T>>());
+    let dataStart = __alloc(byteLength, idof<ArrayBuffer>());
+
+    // create the return value
+    let result: valueof<T>[] = changetype<valueof<T>[]>(__alloc(offsetof<valueof<T>[]>(), idof<valueof<T>[]>()));
+    store<i32>(changetype<usize>(result), size, offsetof<valueof<T>[]>("length_"));
+    result.dataLength = byteLength;
+    result.dataStart = dataStart;
+    // this statement retains the ArrayBuffer
+    result.data = changetype<ArrayBuffer>(dataStart);
+
+    // set the elements
+    let resultIndex: i32 = -1;
+    for (let i = 0; i < length; i++) { // for each child
+      let child: T = load<T>(selfDataStart + (i << alignof<T>()));
+      let childDataLength = child.dataLength;
+      resultIndex++;
+
+      if (child == null) {
+        // if child == null and valueof<T> is a numeric type, it will push 0 here
+        store<valueof<T>>(dataStart + (<usize>resultIndex << usize(alignof<valueof<T>>())), null);
+        continue;
+      }
+
+      let childDataStart = child.dataStart;
+      memory.copy(
+        dataStart + (<usize>resultIndex << usize(alignof<valueof<T>>())),
+        childDataStart,
+        childDataLength,
+      );
+
+      let childLength: i32 = (childDataLength >> usize(alignof<valueof<T>>()));
+      resultIndex += childLength - 1;
+
+      if (isManaged<valueof<T>>()) {
+        for (let j = 0; j < childLength; j++) __retain(load<usize>(childDataStart + (j << usize(alignof<valueof<T>>()))));
```

Comment:
That's interesting approach. We usually split cases for managed copy with retain and loop and unmanaged `memory.copy` but this approach just do optional second walk-through for retaining ref counts.

@dcodeIO wdyt what is approach more performant? This definitely more elegant)

Reply:
Looks good to me. Not sure what's better with current memcpy, but with the bulk instruction this one here might be ideal.

Reply:
Yes, I also thought about bulk operations

Reply:
Thanks guys. Let me know if there is anything that needs to be done.